### PR TITLE
Added spark driver info to UI with alerts

### PR DESCRIPTION
**Pull request by DanielAronovich**
Added spark driver info to the memory usage in the status and summary tabs
Added relevant alerts for driver over usage

### CODE REVIEW

**Review by menishmueli — CHANGES_REQUESTED** [review]

File: spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/listener/model.scala [hunk 1]
```diff
@@ -54,3 +54,13 @@ class IcebergCommitWrapper(val info: IcebergCommitInfo) {
...
+
+case class DataflintEnvironmentInfo(driverXmxBytes: Long)
+
+case class DataflintEnvironmentInfoEvent(environmentInfo: DataflintEnvironmentInfo) extends SparkListenerEvent
+
+class DataflintEnvironmentInfoWrapper(val info: DataflintEnvironmentInfo) {
+  @KVIndex
+  @JsonIgnore
+  def id: String = "environment_info"
+}
```
[the remark this answers is not in the document]
Comment (DanielAronovich) [reply spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/listener/model.scala:66]: You need it as it fails without it, also the IcebergCommitWrapper has it

[the remark this answers is not in the document]
Comment (menishmueli) [reply spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/listener/model.scala:66]: OK

File: spark-ui/src/reducers/Alerts/MemoryAlertsReducer.ts [hunk 2]
```diff
@@ -76,4 +80,29 @@ export function reduceMemoryAlerts(
...
+      name: "driverMemoryTooSmall",
+      title: "Driver Memory Too Small",
+      message: "The Spark driver memory is set to less than 1GB, which may be insufficient for most workloads.",
+      location: `In: Spark Driver (Current llocated memory is: ${humanFileSize(driverMemory)})`,
+      suggestion: `Consider increasing the driver memory to at least 2GB by adding this configuration:
+.config("spark.driver.memory", "2g")`,
+      type: "warning",
+      source: { type: "status", metric: "memory" }
+    });
+  }
```
[the remark this answers is not in the document]
Comment (menishmueli) [reply spark-ui/src/reducers/Alerts/MemoryAlertsReducer.ts]: you did not address this comment

File: spark-ui/src/reducers/Alerts/MemoryAlertsReducer.ts [hunk 3]
```diff
@@ -76,4 +80,29 @@ export function reduceMemoryAlerts(
...
+function reduceDriverMemoryAlerts(
+  statusStore: StatusStore,
+  config: ConfigStore,
+  environmentInfo: any,
+  alerts: Alerts
+) {
+  const driverMemory = environmentInfo?.driverXmxBytes ?? 0;
+  const totalInputBytes = statusStore.executors?.totalInputBytes ?? 0;
+  const numCores = parseInt(config.configs.find(c => c.key === "spark.executor.cores")?.value ?? "1");
+  if (driverMemory < MIN_DRIVER_MEMORY) {
```
[the remark this answers is not in the document]
Comment (menishmueli) [reply spark-ui/src/reducers/Alerts/MemoryAlertsReducer.ts]: please add a "done" next time you address a comment

File: spark-ui/src/components/SummaryBar.tsx [hunk 4]
```diff
@@ -142,6 +144,22 @@ const SummaryBar: FC = (): JSX.Element => {
...
+                <Typography variant="h6" color="inherit" textAlign={"center"}>
+                  {driverMemoryUsageString}
+                </Typography>
+                <Typography variant="subtitle2">
+                This represents the total JVM heap memory allocated to the Spark driver process. The actual available memory for Spark tasks is typically slightly less due to system overhead.
+                </Typography>
+              </React.Fragment>
+            }
+          ></InfoBox>
       </Grid>
```
[the remark this answers is not in the document]
Comment (menishmueli) [reply spark-ui/src/components/SummaryBar.tsx]: please add a "done" comment or a plus icon next time you address a comment

File: spark-ui/src/components/StatusBar.tsx [hunk 5]
```diff
@@ -111,6 +113,22 @@ const StatusBar: FC = (): JSX.Element => {
...
+                <Typography variant="h6" color="inherit" textAlign={"center"}>
+                  {driverMemoryUsageString}
+                </Typography>
+                <Typography variant="subtitle2">
+                This represents the total JVM heap memory allocated to the Spark driver process. The actual available memory for Spark tasks is typically slightly less due to system overhead.
+                </Typography>
+              </React.Fragment>
+            }
+          ></InfoBox>
         </Grid>
```
[the remark this answers is not in the document]
Comment (menishmueli) [reply spark-ui/src/components/StatusBar.tsx]: please add a "done" comment or a plus icon next time you address a comment

File: spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/DataflintSparkUILoader.scala [hunk 6]
```diff
@@ -67,23 +73,24 @@ class DataflintSparkUIInstaller extends Logging {
...
-        else (listener: SparkListenerInterface,
-            queue: String) => context.listenerBus.addToQueue(listener, queue)
-        // DataflintListener currently only relevant for iceberg or databricks support, so no need to add the listener if iceberg support is off
-        addToQueueMethod(new DataflintListener(context.statusStore.store.asInstanceOf[ElementTrackingStore]), "dataflint")
+        else (listener: SparkListenerInterface, queue: String) => context.listenerBus.addToQueue(listener, queue)
+      addToQueueMethod(dataflintListener, "dataflint")
+      context.listenerBus.post(DataflintEnvironmentInfoEvent(environmentInfo))
+      logInfo(s"Posted DataflintEnvironmentInfoEvent with driver memory: $driverXmxBytes bytes")
+      if ((icebergInstalled && icebergEnabled) || isDatabricks) {
         if (isDatabricks) {
```
[the remark this answers is not in the document]
Comment (menishmueli) [reply spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/DataflintSparkUILoader.scala]: You now has if(isDatabricks) twice nested, please fix this

File: spark-ui/src/components/SummaryBar.tsx [hunk 7]
```diff
@@ -124,20 +136,40 @@ const SummaryBar: FC = (): JSX.Element => {
...
+                variant="subtitle1"
+                color="inherit"
+                fontWeight="bold"
+                style={{ marginTop: "16px" }}
+              >
+                Peak Driver Memory:
               </Typography>
-              <Typography variant="subtitle2">
-                "Memory Usage" is the peak memory usage of the most
+              <Typography variant="body1" color="inherit" textAlign={"center"}>
```
Comment (menishmueli) [inline spark-ui/src/components/SummaryBar.tsx:156]: when no data please hide the text

File: spark-ui/src/components/StatusBar.tsx [hunk 8]
```diff
@@ -47,7 +52,15 @@ const StatusBar: FC = (): JSX.Element => {
   );
   const durationText = humanizeTimeDiff(duration(status?.duration), true);
   const numOfQueries = sql?.sqls.length ?? 0;
-
+  const driverExecutor = executors?.find(
+    (exec: SparkExecutorStore) => exec.id === "driver",
+  );
+  const driverMaxMemory = environmentInfo?.driverXmxBytes ?? 1;
+  const driverMemoryUsage = driverExecutor?.memoryUsageBytes ?? 0;
+  const driverMemoryUsagePercentage =
```
Comment (menishmueli) [inline spark-ui/src/components/StatusBar.tsx:60]: please use the calculatePercentage util I wrote

File: spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/DataflintSparkUILoader.scala [hunk 9]
```diff
@@ -67,23 +73,22 @@ class DataflintSparkUIInstaller extends Logging {
...
-        addToQueueMethod(new DataflintListener(context.statusStore.store.asInstanceOf[ElementTrackingStore]), "dataflint")
-        if (isDatabricks) {
-          addToQueueMethod(DataflintDatabricksLiveListener(context.listenerBus), "dataflint")
-        }
-      } catch {
-        case e: Throwable => logWarning("could not add DataFlint Listeners to listener bus", e)
+        else (listener: SparkListenerInterface, queue: String) => context.listenerBus.addToQueue(listener, queue)
+      addToQueueMethod(dataflintListener, "dataflint")
+      context.listenerBus.post(DataflintEnvironmentInfoEvent(environmentInfo))
+      logInfo(s"Posted DataflintEnvironmentInfoEvent max driver memory: ${driverXmxBytes}")
```
Comment (menishmueli) [inline spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/DataflintSparkUILoader.scala:84]: please remove this log line from production

File: spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/DataflintSparkUILoader.scala [hunk 10]
```diff
@@ -67,23 +73,22 @@ class DataflintSparkUIInstaller extends Logging {
...
+      if (isDatabricks) {
+        addToQueueMethod(DataflintDatabricksLiveListener(context.listenerBus), "dataflint")
+        logInfo("Added DataflintDatabricksLiveListener to the Spark context")
```
Comment (menishmueli) [inline spark-plugin/plugin/src/main/scala/org/apache/spark/dataflint/DataflintSparkUILoader.scala:87]: please remove this line from production

File: spark-ui/src/components/SummaryBar.tsx [hunk 11]
```diff
@@ -40,7 +44,15 @@ const SummaryBar: FC = (): JSX.Element => {
   }
 
   const durationText = humanizeTimeDiff(duration(status.duration));
-
+  const driverExecutor = executors
+    ? Object.values(executors).find((exec) => exec.id === "driver")
+    : undefined;
+  const driverMaxMemory = environmentInfo?.driverXmxBytes ?? 1;
+  const driverMemoryUsage = driverExecutor?.memoryUsageBytes ?? 0;
```
Comment (menishmueli) [inline spark-ui/src/components/SummaryBar.tsx:51]: please change memoryUsageBytes to use only on heap memory (+ executors)

**Review by menishmueli — CHANGES_REQUESTED** [review]

**Pull request merged**